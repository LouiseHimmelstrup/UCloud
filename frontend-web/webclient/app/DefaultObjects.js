import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};