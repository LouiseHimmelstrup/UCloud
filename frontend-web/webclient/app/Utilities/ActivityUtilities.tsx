import { buildQueryString } from "./URIUtilities";

export const activityQuery = (offset: number | null, scrollSize: number) => {
    const params = { scrollSize };
    if (offset !== null) params["offset"] = offset;

    return buildQueryString("/activity/browse/user", params);
}

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}` 

export const getLatestActivity = () => undefined;