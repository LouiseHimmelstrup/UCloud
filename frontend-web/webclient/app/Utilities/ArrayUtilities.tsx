export function removeEntry<T>(arr: T[], index: number):T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1)); 
}


type PrimitiveDataTypes = string | number | boolean
/**
 * @param set 
 * @param entry 
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size === set.size;
}