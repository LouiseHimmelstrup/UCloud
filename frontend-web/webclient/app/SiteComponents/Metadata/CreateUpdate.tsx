import * as React from "react";
import { Form, Header, Dropdown, Button } from "semantic-ui-react";
import { licenseOptions, identifierTypes } from "../../DefaultObjects";
import { createRange } from "../../UtilityFunctions";

const newCollaborator = () => ({ name: "", affiliation: "", orcid: "", gnd: "" });
const newGrant = () => ({ id: "" });
const newIdentifier = () => ({ identifier: "", type: "" });
const newSubject = () => ({ term: "", identifier: "" });

export class CreateUpdate extends React.Component<any, any> {
    constructor(props: any) {
        super(props);
        this.state = {
            title: "",
            description: "",
            license: "",
            keywords: [""],
            notes: "",
            collaborators: [newCollaborator()],
            references: [""],
            grants: [newGrant()],
            subjects: [newSubject()],
            identifiers: [newIdentifier()]
        };
        this.setStateEv = this.setStateEv.bind(this);
        this.setStateEvObject = this.setStateEvObject.bind(this);
        this.setStateEvObject = this.setStateEvObject.bind(this);
    }

    onSubmit(e) {
        e.preventDefault();
        // VALIDATION
        // TODO
        // VALIDATION END
    }

    addRow(e, key) {
        e.preventDefault();
        this.setState(() => ({ [key]: this.state[key].concat([""]) }));
    }

    addCollaborator(e) {
        e.preventDefault();
        this.setState(() => ({ collaborators: this.state.collaborators.concat(newCollaborator()) }));
    }

    addSubject(e) {
        e.preventDefault();
        this.setState(() => ({ subjects: this.state.subjects.concat(newSubject()) }));
    }

    addIdentifier(e) {
        e.preventDefault();
        this.setState(() => ({ identifiers: this.state.identifiers.concat(newIdentifier()) }));
    }

    setStateEv(key) {
        return (e, { value }) => {
            let object = {};
            object[key] = value;
            this.setState(() => object);
        };
    }

    setStateEvList(key) {
        return (value, index) => {
            const list = this.state[key];
            console.log(index, key);
            list[index] = value;
            let object = {
                [key]: list
            };
            this.setState(() => object);
        };
    }

    setStateEvObject(key) {
        console.log(key);
        return (value, index, member) => {
            const list = this.state[key];
            console.log(index, key);
            list[index][member] = value;
            let object = {
                [key]: list
            };
            this.setState(() => object);
        };
    }

    render() {
        console.log(this.state);
        return (
            <Form onSubmit={this.onSubmit}>
                <Form.Field required>
                    <label>Title</label>
                    <Form.Input 
                        placeholder="Title" 
                        value={this.state.title} 
                        onChange={this.setStateEv("title")} 
                        required 
                    />
                </Form.Field>
                <Form.Field required>
                    <label>Description</label>
                    <Form.TextArea placeholder="Description" onChange={this.setStateEv("description")} required />
                </Form.Field>
                <Form.Field>
                    <label>License</label>
                    <LicenseDropdown onChange={this.setStateEv("license")} />
                </Form.Field>
                <Form.Field>
                    <label>Keywords</label>
                    <FormFieldList 
                        amount={this.state.keywords} 
                        name="keyword" 
                        setStateEvList={this.setStateEvList("keywords")} 
                    />
                    <Button content="New keyword" onClick={(e) => this.addRow(e, "keywords")} />
                </Form.Field>
                <Form.Field>
                    <label>Notes</label>
                    <Form.TextArea placeholder="Notes..." onChange={this.setStateEv("notes")} />
                </Form.Field>
                <Form.Field>
                    <label>Collaborators</label>
                    <Collaborators 
                        collaborators={this.state.collaborators} 
                        setStateEvObject={this.setStateEvObject("collaborators")} 
                    />
                    <Button content="Add collaborator" onClick={(e) => this.addCollaborator(e)} />
                </Form.Field>
                <Form.Field>
                    <label>References</label>
                    <FormFieldList 
                        name="reference" 
                        amount={this.state.references} 
                        setStateEvList={this.setStateEvList("references")} 
                    />
                    <Button content="Add reference" onClick={(e) => this.addRow(e, "references")} />
                </Form.Field>
                <Form.Field>
                    <label>Subjects</label>
                    <Subjects subjects={this.state.subjects} setStateEvObject={this.setStateEvObject("subjects")} />
                    <Button content="Add subject" onClick={(e) => this.addSubject(e)} />
                </Form.Field>
                <Form.Field>
                    <label>Related Identifiers</label>
                    <RelatedIdentifiers 
                        identifiers={this.state.identifiers} 
                        setStateEvObject={this.setStateEvObject("identifiers")} 
                    />
                    <Button content="Add identifier" onClick={(e) => this.addIdentifier(e)} />
                </Form.Field>
                <Button type="submit" content="Submit" />
            </Form>
        )
    }
}

interface LicenseDropdownProps {
    onChange: (ev, details) => void
}

class LicenseDropdown extends React.PureComponent<LicenseDropdownProps> {
    constructor(props: any) {
        super(props)
    }

    render() {
        return (
            <Form.Dropdown onChange={(e, other) => this.props.onChange(e, other)}
                search
                searchInput={{ type: "string" }}
                selection
                required
                options={licenseOptions}
                placeholder="Select license"
            />
        );
    }
}

const Subjects = ({ subjects, setStateEvObject }) =>
    subjects.map((s, i) =>
        <Form.Group key={i} widths="equal">
            <Form.Input 
                fluid 
                value={s.term} 
                label="Term" 
                required 
                placeholder="Term..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "term")} 
            />
            <Form.Input 
                fluid 
                value={s.identifier} 
                label="Identifier" 
                placeholder="Identifier..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "identifier")} 
            />
        </Form.Group>
    );

const RelatedIdentifiers = ({ identifiers, setStateEvObject }) =>
    identifiers.map((_, i) =>
        <Form.Group key={i} widths="equal">
            <Form.Input 
                fluid label="Identifier" 
                required 
                placeholder="Identifier..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "identifier")} 
            />
            <Form.Dropdown label="Type"
                search
                searchInput={{ type: "string" }}
                selection
                options={identifierTypes}
                placeholder="Select type"
                onChange={(e, { value }) => setStateEvObject(value, i, "type")}
            />
        </Form.Group>
    );


const Collaborators = ({ collaborators, setStateEvObject }) =>
    collaborators.map((c, i) =>
        <Form.Group key={i} widths="equal">
            <Form.Input 
                fluid 
                label="Name" 
                required 
                placeholder="Name..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "name")} 
            />

            <Form.Input 
                fluid 
                label="Affiliation" 
                placeholder="Affiliation..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "affiliation")} 
            />
            
            <Form.Input 
                fluid 
                label="ORCiD" 
                placeholder="ORCiD..." 
                onChange={(e, { value }) => setStateEvObject(value, i, "orcid")} 
            />
            
            <Form.Input 
                fluid 
                label="GND" 
                placeholder="GND" 
                onChange={(e, { value }) => setStateEvObject(value, i, "gnd")} 
            />
        </Form.Group>
    );

const FormFieldList = ({ amount, name, setStateEvList }) =>
    amount.map((c, i) => <Form.Input key={i} value={c} placeholder={`Enter ${name}`} onChange={(e, { value }) => setStateEvList(value, i)} />);