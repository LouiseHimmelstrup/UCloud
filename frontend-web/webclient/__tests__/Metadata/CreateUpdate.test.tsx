import * as React from "react";
import { create } from "react-test-renderer";
import CreateUpdate from "Project/CreateUpdate";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { configureStore } from "Utilities/ReduxUtilities";
import { initStatus } from "DefaultObjects";
import status from "Navigation/Redux/StatusReducer";
import "jest-styled-components";

describe("CreateUpdate Component", () => {
    test.skip("Mount", () => {
        expect(create(
            <Provider store={configureStore({ status: initStatus() }, { status })}>
                <CreateUpdate
                    location={{ search: "?" }}
                    history={createMemoryHistory()}
                />
            </Provider>
        ).toJSON()).toMatchSnapshot()
    });
});