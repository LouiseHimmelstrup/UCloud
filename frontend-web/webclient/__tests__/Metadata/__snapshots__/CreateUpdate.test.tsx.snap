// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateUpdate Component Mount 1`] = `
.c1 {
  cursor: auto;
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
}

.c2 {
  cursor: auto;
  padding-top: 14px;
  padding-right: 14px;
}

.c4 {
  cursor: auto;
  margin-bottom: 1em;
}

.c10 {
  cursor: auto;
}

.c16 {
  cursor: auto;
  margin-top: 8px;
  margin-left: -32px;
}

.c19 {
  cursor: auto;
  margin-bottom: 0.5em;
}

.c24 {
  cursor: auto;
  margin-right: 0.5em;
  width: 25%;
}

.c26 {
  cursor: auto;
  margin-right: 0.5em;
  width: 50%;
}

.c27 {
  cursor: auto;
  width: 50%;
}

.c29 {
  cursor: auto;
  margin-left: auto;
}

.c30 {
  cursor: auto;
  margin-top: 1em;
  margin-bottom: 1em;
}

.c14 {
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 0;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23 {
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  cursor: inherit;
  font-weight: 700;
}

.c7 {
  cursor: inherit;
  color: #c00;
}

.c12 {
  cursor: pointer;
}

.c17 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-bottom: 9px;
}

.c22 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #006aff;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c22:disabled {
  opacity: 0.25;
}

.c22:focus {
  outline: none;
}

.c22:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c25 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #006aff;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
  margin-top: 0.5em;
}

.c25:disabled {
  opacity: 0.25;
}

.c25:focus {
  outline: none;
}

.c25:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c28 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #006aff;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
  margin-bottom: 1em;
}

.c28:disabled {
  opacity: 0.25;
}

.c28:focus {
  outline: none;
}

.c28:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c31 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #00C05A;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c31:disabled {
  opacity: 0.25;
}

.c31:focus {
  outline: none;
}

.c31:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c5 {
  font-size: 10px;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  display: block;
  margin: 0;
  font-size: 14px;
  color: #000;
  font-weight: 700;
  width: 100%;
}

.c8 {
  display: block;
  font-family: inherit;
  color: inherit;
  background-color: transparent;
  border-width: 0px;
  border-style: solid;
  border-color: #c9d3df;
  padding-top: 7px;
  padding-bottom: 7px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #c9d3df;
  box-shadow: 0 0 0 1px #c9d3df;
  border-radius: 5px;
  width: 100%;
}

.c8::-webkit-input-placeholder {
  color: #8393A7;
}

.c8::-moz-placeholder {
  color: #8393A7;
}

.c8:-ms-input-placeholder {
  color: #8393A7;
}

.c8::placeholder {
  color: #8393A7;
}

.c8::-ms-clear {
  display: none;
}

.c8:focus {
  outline: none;
}

.c8:focus {
  outline: 0;
  border-color: #006aff;
  box-shadow: 0 0 0 2px #006aff;
}

.c15 {
  display: block;
  font-family: inherit;
  color: inherit;
  background-color: transparent;
  border-width: 0px;
  border-style: solid;
  border-color: #c9d3df;
  padding-top: 7px;
  padding-bottom: 7px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #c9d3df;
  box-shadow: 0 0 0 1px #c9d3df;
  padding-left: 8px;
  padding-right: 40px;
  border-radius: 5px;
  width: 100%;
}

.c15::-webkit-input-placeholder {
  color: #8393A7;
}

.c15::-moz-placeholder {
  color: #8393A7;
}

.c15:-ms-input-placeholder {
  color: #8393A7;
}

.c15::placeholder {
  color: #8393A7;
}

.c15::-ms-clear {
  display: none;
}

.c15:focus {
  outline: none;
}

.c15:focus {
  outline: 0;
  border-color: #006aff;
  box-shadow: 0 0 0 2px #006aff;
}

.c20 {
  display: block;
  font-family: inherit;
  color: inherit;
  background-color: transparent;
  border-width: 0px;
  border-style: solid;
  border-color: #c9d3df;
  padding-top: 7px;
  padding-bottom: 7px;
  padding-left: 12px;
  padding-right: 12px;
  margin: 0;
  border-color: #c9d3df;
  box-shadow: 0 0 0 1px #c9d3df;
  margin-top: 0.4em;
  margin-bottom: 0.4em;
  border-radius: 5px;
  width: 100%;
}

.c20::-webkit-input-placeholder {
  color: #8393A7;
}

.c20::-moz-placeholder {
  color: #8393A7;
}

.c20:-ms-input-placeholder {
  color: #8393A7;
}

.c20::placeholder {
  color: #8393A7;
}

.c20::-ms-clear {
  display: none;
}

.c20:focus {
  outline: none;
}

.c20:focus {
  outline: 0;
  border-color: #006aff;
  box-shadow: 0 0 0 2px #006aff;
}

.c13 {
  cursor: auto;
}

.c13 .c0 {
  pointer-events: none;
}

.c11 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c11:hover > div {
  display: block;
}

.c18 {
  overflow: visible;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  box-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,.14),0px 1px 8px 0px rgba(0,0,0,.12);
  position: absolute;
  background-color: #fff;
  color: #000;
  width: 100%;
  min-width: 138pxpx;
  padding: 12px 16px;
  z-index: 47;
  text-align: left;
  cursor: pointer;
  opacity: 0;
  pointer-events: none;
}

.c18 > *:hover:not(.c21) {
  background-color: rgba(0,0,0,0.05);
}

.c18 svg {
  margin-right: 1em;
}

.c18 > svg ~ span {
  margin-right: 1em;
}

.c9 {
  width: 100%;
  border-color: #c9d3df;
  box-shadow: 0 0 0 1px #c9d3df;
  border-radius: 5px;
  border: 1px #c9d3df;
  padding: 5px;
  resize: none;
}

.c9:focus {
  outline: 0;
  border-color: #006aff;
  box-shadow: 0 0 0 2px #006aff;
}

.c9:focus {
  outline: none;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c3 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c3 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c3 {
    display: none;
  }
}

<div
  className="c0 c1"
  cursor="auto"
>
  <div
    className="c0 c2"
    cursor="auto"
  >
    <div
      className="c0 c3"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label
          className="c5"
          color="black"
          fontSize={1}
          fontWeight="bold"
          width="100%"
        >
          <span
            className="c6"
            cursor="inherit"
          >
            Title
          </span>
          <span
            className="c7"
            color="red"
            cursor="inherit"
          >
             *
          </span>
          <input
            className="c8"
            id="default"
            onChange={[Function]}
            placeholder="Title"
            required={true}
            type="text"
            value=""
            width="100%"
          />
        </label>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label
          className="c5"
          color="black"
          fontSize={1}
          fontWeight="bold"
          width="100%"
        >
          <span
            className="c6"
            cursor="inherit"
          >
            Description
          </span>
          <span
            className="c7"
            color="red"
            cursor="inherit"
          >
             *
          </span>
          <textarea
            className="c9"
            onChange={[Function]}
            placeholder="Description"
            required={true}
            rows={6}
            value=""
            width={1}
          />
        </label>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label>
          <div
            className="c6"
            cursor="inherit"
          >
            License
          </div>
          <div
            className="c0 c10"
            cursor="auto"
          >
            No license selected
          </div>
          <div
            className="c11"
            data-tag="dropdown"
          >
            <span
              className="c12"
              cursor="pointer"
              onClick={[Function]}
            >
              <div
                className="c0 c13"
                cursor="auto"
              >
                <div
                  className="c14"
                  cursor="inherit"
                  width={1}
                >
                  <input
                    aria-label={null}
                    autoComplete="off"
                    className="c15"
                    id="d"
                    onChange={[Function]}
                    placeholder="Search for a license..."
                    style={
                      Object {
                        "paddingBottom": "9.5px",
                        "paddingTop": "9.5px",
                        "transition": "padding-top 0.1s, padding-bottom 0.1s",
                      }
                    }
                    type="text"
                    value=""
                    width={1}
                  />
                  <div
                    className="c0 c16"
                    cursor="auto"
                  >
                    <svg
                      className="c17"
                      clipRule="evenodd"
                      cursor="inherit"
                      fill="currentcolor"
                      fillRule="evenodd"
                      height={14}
                      mb="9px"
                      viewBox="0 0 24 12"
                      width={14}
                    >
                      <path
                        d="M0 0l11.959 12 12-12H0z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </span>
            <div
              className="c18"
              color="black"
              cursor="pointer"
              disabled={false}
              onClick={[Function]}
              overflow="visible"
              width="100%"
            >
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                BSD Zero Clause License
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Attribution Assurance License
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Abstyles License
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Adobe Systems Incorporated Source Code License Agreement
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Adobe Glyph List License
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Amazon Digital Services License
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Academic Free License v1.1
              </div>
              <div
                className="c0 c19"
                cursor="auto"
                onClick={[Function]}
              >
                Academic Free License v1.2
              </div>
            </div>
          </div>
        </label>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <div
          className="c0 c10"
          cursor="auto"
        >
          <label>
            <span
              className="c6"
              cursor="inherit"
            >
              Keywords
            </span>
            <input
              className="c20"
              id="default"
              onChange={[Function]}
              placeholder="Enter keyword"
              value=""
              width="100%"
            />
          </label>
        </div>
        <button
          className="c21 c22"
          color="blue"
          onClick={[Function]}
          type="button"
        >
          New keyword
        </button>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label>
          Notes
          <textarea
            className="c9"
            onChange={[Function]}
            placeholder="Notes..."
            value=""
            width={1}
          />
        </label>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label>
          Data Management Plan
          <textarea
            className="c9"
            onChange={[Function]}
            placeholder="Data Management Plan..."
            value=""
            width={1}
          />
        </label>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <div
          className="c0 c10"
          cursor="auto"
        >
          <label>
            <span
              className="c6"
              cursor="inherit"
            >
              Contributors
            </span>
            <div
              className="c23"
              cursor="inherit"
            >
              <div
                className="c0 c24"
                cursor="auto"
                width={0.25}
              >
                <label>
                  <span
                    className="c6"
                    cursor="inherit"
                  >
                    Name
                  </span>
                  <input
                    className="c8"
                    id="default"
                    onChange={[Function]}
                    placeholder="Name..."
                    value=""
                    width="100%"
                  />
                </label>
              </div>
              <div
                className="c0 c24"
                cursor="auto"
                width={0.25}
              >
                <label>
                  <span
                    className="c6"
                    cursor="inherit"
                  >
                    Affiliation
                  </span>
                  <input
                    className="c8"
                    id="default"
                    onChange={[Function]}
                    placeholder="Affiliation..."
                    value=""
                    width="100%"
                  />
                </label>
              </div>
              <div
                className="c0 c24"
                cursor="auto"
                width={0.25}
              >
                <label>
                  <span
                    className="c6"
                    cursor="inherit"
                  >
                    ORCID
                  </span>
                  <input
                    className="c8"
                    id="default"
                    onChange={[Function]}
                    placeholder="ORCID..."
                    value=""
                    width="100%"
                  />
                </label>
              </div>
              <div
                className="c0 c24"
                cursor="auto"
                width={0.25}
              >
                <label>
                  <span
                    className="c6"
                    cursor="inherit"
                  >
                    GND
                  </span>
                  <input
                    className="c8"
                    id="default"
                    onChange={[Function]}
                    placeholder="GND..."
                    value=""
                    width="100%"
                  />
                </label>
              </div>
            </div>
          </label>
        </div>
        <button
          className="c21 c25"
          color="blue"
          onClick={[Function]}
          type="button"
        >
          Add collaborator
        </button>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <div
          className="c0 c10"
          cursor="auto"
        >
          <label>
            <span
              className="c6"
              cursor="inherit"
            >
              References
            </span>
            <input
              className="c20"
              id="default"
              onChange={[Function]}
              placeholder="Enter reference"
              value=""
              width="100%"
            />
          </label>
        </div>
        <button
          className="c21 c22"
          color="blue"
          onClick={[Function]}
          type="button"
        >
          Add reference
        </button>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <div
          className="c0 c10"
          cursor="auto"
        >
          <label>
            <span
              className="c6"
              cursor="inherit"
            >
              Grants
            </span>
            <input
              className="c20"
              id="default"
              onChange={[Function]}
              placeholder="Enter grant"
              value=""
              width="100%"
            />
          </label>
        </div>
        <button
          className="c21 c22"
          color="blue"
          onClick={[Function]}
          type="button"
        >
          Add grant
        </button>
      </div>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label>
          <span
            className="c6"
            cursor="inherit"
          >
            Subjects
          </span>
          <div
            className="c23"
            cursor="inherit"
          >
            <div
              className="c0 c26"
              cursor="auto"
              width={0.5}
            >
              <label>
                <span
                  className="c6"
                  cursor="inherit"
                >
                  Term
                </span>
                <input
                  className="c8"
                  id="default"
                  onChange={[Function]}
                  placeholder="Term..."
                  value=""
                  width="100%"
                />
              </label>
            </div>
            <div
              className="c0 c27"
              cursor="auto"
              width={0.5}
            >
              <label>
                <span
                  className="c6"
                  cursor="inherit"
                >
                  Identifier
                </span>
                <input
                  className="c8"
                  id="default"
                  onChange={[Function]}
                  placeholder="Identifier..."
                  value=""
                  width="100%"
                />
              </label>
            </div>
          </div>
        </label>
      </div>
      <button
        className="c21 c28"
        color="blue"
        onClick={[Function]}
        type="button"
      >
        Add subject
      </button>
      <div
        className="c0 c4"
        cursor="auto"
      >
        <label>
          <span
            className="c6"
            cursor="inherit"
          >
            Related identifiers
          </span>
          <div
            className="c23"
            cursor="inherit"
          >
            <div
              className="c0 c26"
              cursor="auto"
              width={0.5}
            >
              <label>
                <span
                  className="c6"
                  cursor="inherit"
                >
                  Identifier
                </span>
                <input
                  className="c8"
                  id="default"
                  onChange={[Function]}
                  placeholder="Identifier..."
                  value=""
                  width="100%"
                />
              </label>
            </div>
            <div
              className="c0 c27"
              cursor="auto"
              width={0.5}
            >
              <label>
                <span
                  className="c6"
                  cursor="inherit"
                >
                  Type
                </span>
                <div
                  className="c11"
                  data-tag="dropdown"
                >
                  <span
                    className="c12"
                    cursor="pointer"
                    onClick={[Function]}
                  >
                    <div
                      className="c0 c13"
                      cursor="auto"
                    >
                      <div
                        className="c14"
                        cursor="inherit"
                        width={1}
                      >
                        <input
                          aria-label={null}
                          autoComplete="off"
                          className="c15"
                          id="d"
                          onChange={[Function]}
                          placeholder="Select type"
                          style={
                            Object {
                              "paddingBottom": "9.5px",
                              "paddingTop": "9.5px",
                              "transition": "padding-top 0.1s, padding-bottom 0.1s",
                            }
                          }
                          type="text"
                          value=""
                          width={1}
                        />
                        <div
                          className="c0 c16"
                          cursor="auto"
                        >
                          <svg
                            className="c17"
                            clipRule="evenodd"
                            cursor="inherit"
                            fill="currentcolor"
                            fillRule="evenodd"
                            height={14}
                            mb="9px"
                            viewBox="0 0 24 12"
                            width={14}
                          >
                            <path
                              d="M0 0l11.959 12 12-12H0z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </span>
                  <div
                    className="c18"
                    color="black"
                    cursor="pointer"
                    disabled={false}
                    onClick={[Function]}
                    overflow="visible"
                    width="100%"
                  >
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Cited by
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Cites
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Supplement to
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Supplemented by
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      New version of
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Previous version of
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Part of
                    </div>
                    <div
                      className="c0 c19"
                      cursor="auto"
                      onClick={[Function]}
                    >
                      Has part
                    </div>
                  </div>
                </div>
              </label>
            </div>
          </div>
        </label>
      </div>
      <button
        className="c21 c22"
        color="blue"
        onClick={[Function]}
        type="button"
      >
        Add identifier
      </button>
      <div
        className="c23"
        cursor="inherit"
      >
        <div
          className="c0 c29"
          cursor="auto"
        />
        <div
          className="c0 c30"
          cursor="auto"
        >
          <button
            className="c21 c31"
            color="green"
            onClick={[Function]}
            type="button"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
